### PR TITLE
[Feat] Distrifusion Acceleration Support for Diffusion Inference

### PR DESCRIPTION
**Pull request by LRY89757**
## 📌 Checklist before creating the PR

- [ ] I have created an issue for this PR for traceability
- [x] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [ ] I have added relevant tags if possible for us to better distinguish different PRs
- [ ] I have installed pre-commit: `pip install pre-commit && pre-commit install`

## 📝 What does this PR do?

1. Support Distrifusion, CVPR 2024: https://arxiv.org/abs/2402.19481
2. Add benchmark
3. Add FID Test

## 💥 Checklist before requesting a review

- [ ] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [ ] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [x] I have performed a self-review of my code
- [ ] I have added thorough tests.
- [ ] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [x] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

**Comment by LRY89757** [comment]
A Simple Example:

img_v3_02ck_bdccbf5b-b375-4013-b70a-4b8c73bf099g

File: examples/inference/stable_diffusion/sd3_generation.py [hunk 1]
```diff
@@ -51,12 +54,17 @@ def infer(args):
...
     out = engine.generate(prompts=[args.prompt], generation_config=DiffusionGenerationConfig())[0]
-    out.save("cat.jpg")
+    if dist.get_rank() == 0:
+        out.save(f"cat_parallel_size{dist.get_world_size()}.jpg")
     coordinator.print_on_master(out)
 
 
 # colossalai run --nproc_per_node 1 examples/inference/stable_diffusion/sd3_generation.py -m MODEL_PATH
-# colossalai run --nproc_per_node 1 examples/inference/stable_diffusion/sd3_generation.py -m "stabilityai/stable-diffusion-3-medium-diffusers" --tp_size 1
+# CUDA_VISIBLE_DEVICES_set_n_least_memory_usage 1 && colossalai run --nproc_per_node 1 examples/inference/stable_diffusion/sd3_generation.py -m "stabilityai/stable-diffusion-3-medium-diffusers" --tp_size 1
```
Comment (yuanheng-zhao) [inline examples/inference/stable_diffusion/sd3_generation.py:63]: `CUDA_VISIBLE_DEVICES_set_n_least_memory_usage` is only set and exposed in `examples/inference/llama/run_benchmark.sh` and should not be brought as an instruction command

File: colossalai/inference/modeling/layers/distrifusion.py [hunk 2]
```diff
@@ -0,0 +1,640 @@
...
+    )
+
+    # enable Distrifusion Optimization
+    if hasattr(self, "patched_parallel_size"):
+        from torch import distributed as dist
+
+        if getattr(self, "output_buffer", None) is None:
+            self.output_buffer = torch.empty_like(output)
+        if getattr(self, "buffer_list", None) is None:
+            self.buffer_list = [torch.empty_like(output) for _ in range(self.patched_parallel_size)]
```
[the remark this answers is not in the document]
Comment (LRY89757) [reply colossalai/inference/modeling/layers/distrifusion.py:144]: good catch! A possible solution:
Set Max_Image_Height/Width and allocate a buffer that is big enough in advance.

File: colossalai/inference/modeling/layers/distrifusion.py [hunk 3]
```diff
@@ -0,0 +1,640 @@
...
+
+
+# Code adapted from: https://github.com/PipeFusion/PipeFusion/blob/main/pipefuser/modules/dit/patch_parallel/attn.py
+class DistriSelfAttention(ParallelModule):
+    def __init__(
+        self,
+        module: Attention,
+        process_group: Union[ProcessGroup, List[ProcessGroup]],
+        model_shard_infer_config: ModelShardInferenceConfig = None,
+        async_nccl_stream: torch.cuda.Stream = None,
```
[the remark this answers is not in the document]
Comment (LRY89757) [reply colossalai/inference/modeling/layers/distrifusion.py:495]: But there are many attn layers, we want to share the same stream between all the attn layers

File: colossalai/inference/modeling/layers/distrifusion.py [hunk 4]
```diff
@@ -0,0 +1,640 @@
...
+    def __init__(
+        self,
+        module: PatchEmbed,
+        process_group: Union[ProcessGroup, List[ProcessGroup]],
+        model_shard_infer_config: ModelShardInferenceConfig = None,
+    ):
+        super().__init__()
+        self.module = module
+        self.rank = dist.get_rank(group=process_group)
+        self.patched_parallelism_size = model_shard_infer_config.patched_parallelism_size
```
[the remark this answers is not in the document]
Comment (LRY89757) [reply colossalai/inference/modeling/layers/distrifusion.py:222]: We discussed it, dist.process_group will init after we get patched_parallelism_size from  inferenceconfig. so here we choose to use it

File: colossalai/inference/modeling/layers/distrifusion.py [hunk 5]
```diff
@@ -0,0 +1,640 @@
...
+            attention_mask=attention_mask,
+            encoder_hidden_states=encoder_hidden_states,
+            encoder_attention_mask=encoder_attention_mask,
+            timestep=timestep,
+            cross_attention_kwargs=cross_attention_kwargs,
+            class_labels=class_labels,
+        )
+
+    # 3. Output
+    shift, scale = (self.scale_shift_table[None] + embedded_timestep[:, None].to(self.scale_shift_table.device)).chunk(
```
[the remark this answers is not in the document]
Comment (LRY89757) [reply colossalai/inference/modeling/layers/distrifusion.py:107]: Yes, they're in the same process, same current_device

**Review by yuanheng-zhao — APPROVED** [review]
LGTM

**Comment by LRY89757** [comment]
Performance:

PixArtAlpha
image
StableDiffusion 3
image

**Pull request merged**